File: src/main/java/org/team1/nbe1_3_team01/domain/attendance/controller/AttendanceController.java
package org.team1.nbe1_3_team01.domain.attendance.controller;

import jakarta.validation.Valid;
import java.net.URI;
import java.text.MessageFormat;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.team1.nbe1_3_team01.domain.attendance.service.dto.AttendanceCreateRequest;
import org.team1.nbe1_3_team01.domain.attendance.service.dto.AttendanceIdRequest;
import org.team1.nbe1_3_team01.domain.attendance.service.dto.AttendanceUpdateRequest;
import org.team1.nbe1_3_team01.domain.attendance.service.AttendanceQueryService;
import org.team1.nbe1_3_team01.domain.attendance.service.AttendanceCommandService;
import org.team1.nbe1_3_team01.domain.attendance.service.response.AttendanceIdResponse;
import org.team1.nbe1_3_team01.domain.attendance.controller.response.AttendanceResponse;
import org.team1.nbe1_3_team01.global.util.Response;
import org.team1.nbe1_3_team01.global.util.SecurityUtil;

@RestController
@RequestMapping("/api/attendances")
@RequiredArgsConstructor
public class AttendanceController {

    private final AttendanceCommandService attendanceCommandService;
    private final AttendanceQueryService attendanceQueryService;

    /**
     * 자신의 출결 요청 보기
     */
    @GetMapping
    public ResponseEntity<Response<List<AttendanceResponse>>> getMyAttendances() {
        var currentUsername = SecurityUtil.getCurrentUsername();

        return ResponseEntity.ok(
                Response.success(attendanceQueryService.getMyAttendances(currentUsername)));
    }

    /**
     * 자신의 출결 요청 상세 보기
     */
    @GetMapping("/{id}")
    public ResponseEntity<Response<AttendanceResponse>> getMyAttendance(
            @PathVariable("id") Long attendanceId
    ) {
        var currentUsername = SecurityUtil.getCurrentUsername();

        return ResponseEntity.ok(
                Response.success(attendanceQueryService.getByIdAndUserId(attendanceId, currentUsername)));
    }

    /**
     * 출결 요청 등록
     */
    @PostMapping
    public ResponseEntity<Response<AttendanceIdResponse>> registAttendance(
            @RequestBody @Valid AttendanceCreateRequest attendanceCreateRequest
    ) {
        var registerUsername = SecurityUtil.getCurrentUsername();

        var attendanceIdResponse = attendanceCommandService.register(registerUsername, attendanceCreateRequest);
        return ResponseEntity
                .created(URI.create(MessageFormat.format("/api/attendances/{0}", attendanceIdResponse.attendanceId())))
                .body(Response.success(attendanceIdResponse));
    }

    /**
     * 출결 요청 수정
     */
    @PatchMapping
    public ResponseEntity<Response<AttendanceIdResponse>> updateAttendance(
            @RequestBody @Valid AttendanceUpdateRequest attendanceUpdateRequest
    ) {
        var currentUsername = SecurityUtil.getCurrentUsername();

        return ResponseEntity.ok(
                Response.success(attendanceCommandService.update(currentUsername, attendanceUpdateRequest)));
    }

    /**
     * 출결 요청 삭제
     */
    @DeleteMapping
    public ResponseEntity<Void> deleteAttendance(
            @RequestBody AttendanceIdRequest attendanceIdRequest
    ) {
        var currentUsername = SecurityUtil.getCurrentUsername();

        attendanceCommandService.delete(currentUsername, attendanceIdRequest.id());
        return ResponseEntity.noContent()
                .build();
    }
}

File: src/main/java/org/team1/nbe1_3_team01/domain/attendance/controller/AttendanceController.kt
package org.team1.nbe1_3_team01.domain.attendance.controller

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.team1.nbe1_3_team01.domain.attendance.service.AttendanceCommandService
import org.team1.nbe1_3_team01.domain.attendance.service.AttendanceQueryService
import org.team1.nbe1_3_team01.domain.attendance.service.dto.AttendanceCreateRequest
import org.team1.nbe1_3_team01.domain.attendance.service.dto.AttendanceIdRequest
import org.team1.nbe1_3_team01.domain.attendance.service.dto.AttendanceUpdateRequest
import org.team1.nbe1_3_team01.domain.attendance.service.response.AttendanceIdResponse
import org.team1.nbe1_3_team01.domain.attendance.controller.response.AttendanceResponse
import org.team1.nbe1_3_team01.global.util.Response
import org.team1.nbe1_3_team01.global.util.SecurityUtil
import java.net.URI
import jakarta.validation.Valid

@RestController
@RequestMapping("/api/attendances")
class AttendanceController(
    private val attendanceCommandService: AttendanceCommandService,
    private val attendanceQueryService: AttendanceQueryService
) {

    /**
     * 자신의 출결 요청 보기
     */
    @GetMapping
    fun getMyAttendances(): ResponseEntity<Response<List<AttendanceResponse>>> {
        val currentUsername = SecurityUtil.getCurrentUsername()

        val attendances = attendanceQueryService.getMyAttendances(currentUsername)
        return ResponseEntity.ok(Response.success(attendances))
    }

    /**
     * 자신의 출결 요청 상세 보기
     */
    @GetMapping("/{id}")
    fun getMyAttendance(
        @PathVariable("id") attendanceId: Long
    ): ResponseEntity<Response<AttendanceResponse>> {
        val currentUsername = SecurityUtil.getCurrentUsername()

        val attendance = attendanceQueryService.getByIdAndUserId(attendanceId, currentUsername)
        return ResponseEntity.ok(Response.success(attendance))
    }

    /**
     * 출결 요청 등록
     */
    @PostMapping
    fun registAttendance(
        @RequestBody @Valid attendanceCreateRequest: AttendanceCreateRequest
    ): ResponseEntity<Response<AttendanceIdResponse>> {
        val registerUsername = SecurityUtil.getCurrentUsername()

        val attendanceIdResponse = attendanceCommandService.register(registerUsername, attendanceCreateRequest)

        val location = URI.create("/api/attendances/${attendanceIdResponse.attendanceId}")
        return ResponseEntity.created(location).body(Response.success(attendanceIdResponse))
    }

    /**
     * 출결 요청 수정
     */
    @PatchMapping
    fun updateAttendance(
        @RequestBody @Valid attendanceUpdateRequest: AttendanceUpdateRequest
    ): ResponseEntity<Response<AttendanceIdResponse>> {
        val currentUsername = SecurityUtil.getCurrentUsername()

        val updatedAttendance = attendanceCommandService.update(currentUsername, attendanceUpdateRequest)
        return ResponseEntity.ok(Response.success(updatedAttendance))
    }

    /**
     * 출결 요청 삭제
     */
    @DeleteMapping
    fun deleteAttendance(
        @RequestBody attendanceIdRequest: AttendanceIdRequest
    ): ResponseEntity<Void> {
        val currentUsername = SecurityUtil.getCurrentUsername()

        attendanceCommandService.delete(currentUsername, attendanceIdRequest.id)
        return ResponseEntity.noContent().build()
    }
}
